api/app/Models/PoolCandidate.php

```diff
@@ -911,4 +919,162 @@ public function setApplicationSnapshot()
         $this->profile_snapshot = $profile;
         $this->save();
     }
+
+    public function computeAssessmentStatus()
+    {
+        $decisions = [];
+        $currentStep = 1;
+        $this->load([
+            'pool.assessmentSteps',
+            'pool.assessmentSteps.poolSkills',
+            'assessmentResults',
+            'assessmentResults.poolSkill',
+        ]);
+
+        foreach ($this->pool->assessmentSteps as $step) {
+            $stepId = $step->id;
+            $hasFailure = false;
+            $hasOnHold = false;
+            $hasToAssess = false;
+
+            $stepResults = $this->assessmentResults->where('assessment_step_id', $stepId);
+            if ($stepResults->isEmpty()) {
+                continue;
+            }
+
+            $hasEssentialSkillsToAssess = $step->poolSkills->contains(function ($poolSkill) {
+                return $poolSkill->type === PoolSkillType::ESSENTIAL->name;
+            });
+
+            if ($hasEssentialSkillsToAssess) {
```

Comment:
So what happens if an assessment step has no essential skills (and doesn't require education)? See my earlier comment about that test case. Perhaps double check with Jerbo on what the desired behaviour is in that case.